src/osd/PrimaryLogPG.h

```diff
@@ -1622,18 +1622,25 @@ class PrimaryLogPG : public PG, public PGBackend::Listener {
       > reactions;
     struct ReservationCB : public Context {
       PrimaryLogPGRef pg;
+      OSDService *service;
       bool canceled;
-      ReservationCB(PrimaryLogPG *pg) : pg(pg), canceled(false) {}
+      ReservationCB(PrimaryLogPG *pg) : pg(pg), service(pg->osd),
+					canceled(false) {}
       void finish(int) override {
-	pg->lock();
+	PrimaryLogPG *pgp = pg.get();
+	if (service->is_stopping() || !pgp)
+	  return;
+	pgp->lock();
 	if (!canceled)
-	  pg->snap_trimmer_machine.process_event(SnapTrimReserved());
-	pg->unlock();
+	  pgp->snap_trimmer_machine.process_event(SnapTrimReserved());
+	pgp->unlock();
       }
       void cancel() {
 	assert(pg->is_locked());
 	assert(!canceled);
 	canceled = true;
+	pg->osd->snap_reserver.cancel_reservation(pg->get_pgid());
+	pg.reset();
```

Comment:
`pg` will call `pg->put("intptr")` in its dtor, so it's not necessary to reset it explicitly, am i right?